### PR TITLE
Allow authentication with registry.npmjs.com

### PR DESCRIPTION
**Pull request by Ianfeather**
This adds functionality for https://github.com/jdxcode/npm-register/issues/93

Making `config.uplink` an object and including auth as a property would be a little nicer imo, but it would be a breaking change and I'm not sure this warrants that change. 

Could you give me some pointers on how to run the tests? Do I need to run them within the container? Currently when I run them locally they're almost all returned as `pending`. Let me know and i'll add tests for this functionality.

cc @dgautsch

### CODE REVIEW

**Comment by dgautsch** [comment]
Looks good so far, can you add update the unit tests? @jdxcode does the package.json need to be updated? Doesn't npm publish handle that?

**Comment by Ianfeather** [comment]
Hey @dgautsch, thanks for getting back to me. Regarding the unit tests:

> Could you give me some pointers on how to run the tests? Do I need to run them within the container? Currently when I run them locally they're almost all returned as pending. Let me know and i'll add tests for this functionality.

I'm more than happy to help fix up existing tests too, but at the moment I have a totally unstable set of tests locally so I assume I'm running them incorrectly. Thanks.

**Comment by jdxcode** [comment]
I definitely wouldn't change the version in a PR. I recommend using `npm version` or the `np` tool to do that when you publish a release

**Comment by Ianfeather** [comment]
@dgautsch @jdxcode If you could tell me how I can run the tests locally I can add unit tests and land this patch?

**Comment by dgautsch** [comment]
@Ianfeather hi Ian, apologies for the slow response, I was hopping jd would jump in here because the s3 tests are going to fail. looks like they're running through his personal amazon cloud. Running tests is done from the command line via `npm test`

**Comment by jdxcode** [comment]
running locally you just need to set the AWS env vars to a valid S3 bucket. Running on circle is a different story, however. I'm not sure how to enable the public to run the tests without potentially exposing the credentials

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "npm-register",
   "description": "Your own private npm registry and backup server",
-  "version": "2.5.2",
+  "version": "2.6.0",
```
Comment (dgautsch) [inline package.json]: @Ianfeather revert this file, I will make sure it's updated when we republish.

Comment (Ianfeather) [reply package.json]: Will do

**Comment by Ianfeather** [comment]
@jdxcode Thanks, that should give me enough to get going.

Would you be interested in an update which mocks s3 for the unit tests? You could potentially have a couple of integration tests that run separately to the core unit tests and actually hit an s3 bucket

**Comment by jdxcode** [comment]
that sounds like a good idea, I wonder if there is something like nock for s3 that would be useful here

**Comment by Ianfeather** [comment]
There's this https://www.npmjs.com/package/mock-aws-s3

I'll circle back to that in a different PR

**Comment by Ianfeather** [comment]
I added test documentation in https://github.com/jdxcode/npm-register/pull/95

**Comment by Ianfeather** [comment]
@dgautsch This now has tests and all are passing locally.

There are commits that are in this and also in #95 - I'd recommend merging 95 if you can first, then i'll rebase this branch. I think they're both good to land though.

Let me know what you think from here.

**Comment by dgautsch** [comment]
Ok thanks Ian. I'll take a look tonight after work.

**Comment by Ianfeather** [comment]
I think this might finally be ready to go! Integration tests pass locally

**Comment by Ianfeather** [comment]
@dgautsch @jdxcode Are you ok to land this patch now?

**Comment by jdxcode** [comment]
👍

**Comment by Ianfeather** [comment]
@dgautsch Sorry to bug you. Are you happy to merge and publish this new version?

**Comment by dgautsch** [comment]
Hey I'll publish, I think I still need npm rights. If I do, I'll do it
today!

On Jan 8, 2018 6:48 PM, "Ianfeather" <notifications@github.com> wrote:

@dgautsch <https://github.com/dgautsch> Sorry to bug you. Are you happy to
merge and publish this new version?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/jdxcode/npm-register/pull/94#issuecomment-356133570>,
or mute the thread

**Comment by jdxcode** [comment]
you should have it @dgautsch!

**Comment by Ianfeather** [comment]
Fantastic, thanks both of you

**Pull request merged**